processor/memorylimiter/memorylimiter.go

```diff
@@ -238,44 +239,89 @@ func (ml *memoryLimiter) forcingDrop() bool {
 	return atomic.LoadInt64(&ml.forceDrop) != 0
 }
 
-func (ml *memoryLimiter) memCheck() {
-	ms := ml.readMemStats()
-	ml.memLimiting(ms)
+func (ml *memoryLimiter) setForcingDrop(b bool) {
+	var i int64
+	if b {
+		i = 1
+	} else {
+		i = 0
+	}
+	atomic.StoreInt64(&ml.forceDrop, i)
 }
 
-func (ml *memoryLimiter) memLimiting(ms *runtime.MemStats) {
-	if !ml.decision.shouldDrop(ms) {
-		atomic.StoreInt64(&ml.forceDrop, 0)
-	} else {
-		atomic.StoreInt64(&ml.forceDrop, 1)
-		// Force a GC at this point and see if this is enough to get to
-		// the desired level.
+func bytesToMib(ms *runtime.MemStats) zap.Field {
+	return zap.Uint64("cur_mem_mib", ms.Alloc/1024/1024)
+}
+
+func (ml *memoryLimiter) checkMemLimits() {
+	ms := ml.readMemStats()
+
+	ml.logger.Debug("Currently used memory.", bytesToMib(ms))
+
+	if ml.usageChecker.aboveHardLimit(ms) {
+		// Memory usage is above hard limit. Force GC.
+		ml.logger.Warn("Memory usage is above hard limit, force GC.", bytesToMib(ms))
 		runtime.GC()
+		ms = ml.readMemStats()
+		ml.logger.Info("Memory usage after GC.", bytesToMib(ms))
+	}
+
+	if ml.forcingDrop() {
+		// Currently forcing drop due to memory limit. See if we still need to do it.
+
+		if ml.usageChecker.aboveSoftLimit(ms) {
+			// Continue dropping, we are still above soft limit.
+			return
+		}
+
+		// Enough memory is available now, no need to limit.
+		ml.logger.Info("Memory usage back within limits, resuming normal operation.", bytesToMib(ms))
+		ml.setForcingDrop(false)
+
+	} else {
+		// Currently not forcing drop. Check if we are still below soft memory limit.
+
+		if !ml.usageChecker.aboveSoftLimit(ms) {
+			// We are still below limits, nothing to do.
+			return
+		}
+
+		// We are above soft limit, so start dropping.
+		ml.setForcingDrop(true)
+		ml.logger.Warn("Memory usage is above soft limit, begin dropping data.", bytesToMib(ms))
 	}
 }
 
-type dropDecision struct {
+type memUsageChecker struct {
 	memAllocLimit uint64
 	memSpikeLimit uint64
 }
 
-func (d dropDecision) shouldDrop(ms *runtime.MemStats) bool {
-	return d.memAllocLimit <= ms.Alloc || d.memAllocLimit-ms.Alloc <= d.memSpikeLimit
+func (d memUsageChecker) aboveSoftLimit(ms *runtime.MemStats) bool {
+	return ms.Alloc >= d.memAllocLimit-d.memSpikeLimit
 }
 
-func newFixedDecision(memAllocLimit, memSpikeLimit uint64) (*dropDecision, error) {
+func (d memUsageChecker) aboveHardLimit(ms *runtime.MemStats) bool {
+	return ms.Alloc >= d.memAllocLimit
+}
+
+func newFixedMemUsageChecker(memAllocLimit, memSpikeLimit uint64) (*memUsageChecker, error) {
 	if memSpikeLimit >= memAllocLimit {
 		return nil, errMemSpikeLimitOutOfRange
 	}
-	return &dropDecision{
+	if memSpikeLimit == 0 {
```

Comment:
People looking at the metrics would thank you if you store this value (and the max) as a gauge (or expvar). This way, they wouldn't need to figure out where the config for this is located in their internal source repositories and which value should currently be used. This is specially true for values that aren't explicit, like this one here.

Reply:
Not sure I understand the "This way, they wouldn't need to figure out where the config for this is located in their internal source repositories and which value should currently be used."? I understand the reasoning to expose this as metric or log, or even in zpages, but not sure I understand the part of looking into source code.

Reply:
Sorry, should have been clearer. People store configuration files in configuration management tools and/or source control mechanisms (git). The person looking at the memory consumption for the collector, perhaps via Prometheus or something even farther away, like a Grafana dashboard, might not readily have access to the configuration files for the collector. So, providing the configuration value as expvar/gauge would make their lives easier.

Reply:
We log the value of memSpikeLimit. A metric can be added in a future PR if we feel strong about it.